tree/treeplayer/src/TTreeReaderValue.cxx

```diff
@@ -133,9 +133,67 @@ void ROOT::Internal::TTreeReaderValueBase::RegisterWithTreeReader() {
 /// Try to read the value from the TBranchProxy, returns
 /// the status of the read.
 
+
+
+template <ROOT::Internal::TTreeReaderValueBase::BranchProxyRead_t Func>
+ROOT::Internal::TTreeReaderValueBase::EReadStatus ROOT::Internal::TTreeReaderValueBase::ProxyReadTemplate()
+{
+   if ((fProxy->*Func)()) {
+      fReadStatus = kReadSuccess;
+   } else {
+      fReadStatus = kReadError;
+   }
+   return fReadStatus;
+}
+
 ROOT::Internal::TTreeReaderValueBase::EReadStatus
-ROOT::Internal::TTreeReaderValueBase::ProxyRead() {
+ROOT::Internal::TTreeReaderValueBase::ProxyReadDefaultImpl() {
    if (!fProxy) return kReadNothingYet;
+   if (fProxy->IsInitialized() || fProxy->Setup()) {
+
+      using EReadType = ROOT::Detail::TBranchProxy::EReadType;
+      using TBranchPoxy = ROOT::Detail::TBranchProxy;
+
+      EReadType readtype = EReadType::kNoDirector;
+      if (fProxy) readtype = fProxy->GetReadType();
+
+      switch (readtype) {
```

Comment:
Would `Setup()` be a more obvious place for setting up `fProxyReadFunc`?

Reply:
Maybe but it is not obvious ...if by Setup you mean moving to TBranchProsy then actually no because Read is an inline function so for other client of Read having read itself calls a function pointer would be pessimissation.

If you mean 'the' routine that setups the TTreeReaderValues ... well at the moment there is not a clear one (NotifyNewTree is not always called as far as I can tell).

So maybe at a later date ...

Reply:
If the `EReadType` starts from 0, and increases one by one, you could create a table of pointers and replace this switch with a table lookup, which would probably be faster.

Reply:
The execution of this code is 'rare' (i.e. only on the first entry of each file the TTreeReaderValue sees).   So I am not sure the change is worth it.